miden-tx/src/executor/mod.rs

```diff
@@ -0,0 +1,125 @@
+    /// Errors:
+    /// - If required data can not be fetched from the [DataStore].
+    /// - If the transaction can not be compiled.
+    fn prepare_transaction(
+        &mut self,
+        account_id: AccountId,
+        block_ref: u32,
+        note_origins: Vec<(u32, NodeIndex)>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<PreparedTransaction, TransactionExecutorError> {
+        let account = self
+            .data_store
+            .get_account(account_id)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let block_header = self
+            .data_store
+            .get_block_header(block_ref)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let notes = note_origins
+            .into_iter()
+            .map(|(block_num, note_index)| {
+                self.data_store
+                    .get_note(block_num, note_index)
+                    .map_err(TransactionExecutorError::FetchTransactionDataFailed)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let block_chain = self
+            .data_store
+            .get_chain_mmr(block_ref)
+            .map_err(TransactionExecutorError::FetchTransactionDataFailed)?;
+
+        let (tx_script_root, tx_program) = self
+            .compiler
+            .compile_transaction(account_id, &notes, tx_script)
+            .map_err(TransactionExecutorError::CompileTransactionError)?;
+
+        Ok(PreparedTransaction::new(
+            account,
+            block_header,
+            block_chain,
+            notes,
+            tx_script_root,
+            tx_program,
+        ))
+    }
+
+    /// Constructs a transaction from the provided inputs, executes it, and returns a [TxWitness]
+    /// which can be used to generate a proof for the transaction.
+    pub fn execute_transaction(
+        &mut self,
+        account_id: AccountId,
+        block_ref: u32,
+        note_origins: Vec<(u32, NodeIndex)>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<TxWitness, TransactionExecutorError> {
+        let transaction =
+            self.prepare_transaction(account_id, block_ref, note_origins, tx_script)?;
+
+        let advice_recorder: AdviceRecorder = transaction.advice_provider_inputs().into();
+        // TODO: Process is currently behind the `internals` feature flag
+        let mut process = Process::new(
+            transaction.tx_program().kernel().clone(),
+            transaction.stack_inputs(),
+            advice_recorder,
+        );
+        let _result = process
+            .execute(transaction.tx_program())
+            .map_err(TransactionExecutorError::ExecuteTransactionProgramFailed)?;
+        let (_, _, _, _, _, advice_recorder) = process.into_parts();
+
+        Ok(TxWitness {
+            program: transaction.tx_program().clone(),
+            stack_witness: transaction.stack_inputs(),
+            advice_witness: advice_recorder.into_proof(),
```

Comment:
I wonder if we should track both stack inputs and stack outputs here (rather than just stack inputs).

Reply:
My thinking was the witness should only include inputs.  The stack outputs can be produced by executing the transaction again.  Is there some cases where we would be consuming the transaction witness, specifically the stack outputs, without executing the transaction again?

Reply:
No specific use case. My thinking was that it may be possible to perform some actions when outputs are also known without (or before) re-executing the program, but that could open up some attack vectors. So, let's keep it as is - and if we find a good use case, we can always add the outputs here too.